<h1><center>Regex正則表達式</center></h1>
●Python 中的正則表達式可通過 re 模塊使用

●使用正則表達式，您可以為要匹配的可能字符串集指定規則。通常我們首先定義我們想要搜索的模式，並re.compile()在其上使用。默認情況下，我們的模式區分大小寫。

<h1><center>注意：建議使用原始字符串進行搜索</center></h1>

In [10]:
## Use raw strings for the search pattern
a = '\tHello'
b = r'\tHello'
print(a)
print(b)

	Hello
\tHello


In [14]:
import re

test_string="123river45678river123river"

# 分開寫 pattern.finditer()/match()/search()/findall()

pattern = re.compile(r"river")
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

print("------------------------------------")

# 寫一起
matches2 = re.finditer(r"123",test_string)
for match2 in matches2:
    print(match2)


<re.Match object; span=(3, 8), match='river'>
<re.Match object; span=(13, 18), match='river'>
<re.Match object; span=(21, 26), match='river'>
------------------------------------
<re.Match object; span=(0, 3), match='123'>
<re.Match object; span=(18, 21), match='123'>


<h1><center>與編譯對象執行匹配</center></h1>
一旦我們有了我們的模式，我們就可以在我們想要查找的文本/字符串中搜索這個模式。

● match()：確定正則是否在字符串的開頭匹配。

● search()：掃描字符串，查找此 RE 匹配的任何位置。

● findall()：查找RE匹配的所有子字符串，並將它們作為列表返回。

● finditer()：查找所有正則匹配的子串，並將它們作為迭代器返回。

In [15]:
# finditer()
my_string = 'river123RIVER123river'
pattern = re.compile(r'123')
matches = pattern.finditer(my_string)
for match in matches:
    print(match)
    print(match.span(), match.start(), match.end())
    print(match.group()) # returns the string

print()
# findall()
pattern = re.compile(r'123')
matches = pattern.findall(my_string)
for match in matches:
    print(match)

print()
# match
match = pattern.match(my_string)
print(match)
pattern = re.compile(r'river')
match = pattern.match(my_string)
print(match)

print()
# search
match = pattern.search(my_string)
print(match)

<re.Match object; span=(5, 8), match='123'>
(5, 8) 5 8
123
<re.Match object; span=(13, 16), match='123'>
(13, 16) 13 16
123

123
123

None
<re.Match object; span=(0, 5), match='river'>

<re.Match object; span=(0, 5), match='river'>


<h1><center>Match 對像上的方法</center></h1>
● group()：返回正則匹配的字符串

● start()：返回匹配的起始位置

● end()：返回匹配的結束位置

● span()：返回一個包含匹配（開始，結束）位置的元組

In [12]:
test_string = '123abc456789abc123ABC'
pattern = re.compile(r'abc')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
    print(match.span(), match.start(), match.end())
    print(match.group()) # returns the substring that was matched by the RE

<re.Match object; span=(3, 6), match='abc'>
(3, 6) 3 6
abc
<re.Match object; span=(12, 15), match='abc'>
(12, 15) 12 15
abc


<h1><center>元字符</center></h1>
元字符是具有特殊含義的

字符：所有元字符：. ^ $ * + ? { } [ ] \ | ( )

如果我們真的要搜索字符，則需要對元字符進行轉義

● . 任何字符（換行符除外）“he..o”

● ^ 以“^hello”開頭

● $ 以“world$”結尾

● * 零次或多次出現“aix*”

● + 一次或多次出現“aix+”

● { } 恰好是指定的出現次數“al{2}”

● [] 一組字符“[am]”

● 表示特殊序列（也可用於轉義特殊字符）“\d”

● 要馬“A|B”

● ( ) 捕獲和分組

In [13]:
test_string = 'python-engineer.com'
pattern = re.compile(r'\.')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

<re.Match object; span=(15, 16), match='.'>


<h1><center>更多元字符/特殊序列</center></h1>

特殊序列是 \ 後跟下表中的字符之一，具有特殊含義：

● \d ：匹配任何十進制數字；這相當於類 [0-9]。

● \D ：匹配任何非數字字符；這相當於類 [^0-9]。

● \s ：匹配任何空白字符；

● \S ：匹配任何非空白字符；

● \w ：匹配任何字母數字（單詞）字符；這相當於 [a-zA-Z0-9_] 類。

● \W ：匹配任何非字母數字字符；這相當於類 [^a-zA-Z0-9_]。

● \b 返回指定字符位於單詞開頭或結尾的匹配項 r"\bain" r"ain\b"

● \B 返回存在指定字符的匹配項，但不在單詞 r"\Bain" r"ain\B" 的開頭（或結尾）

● \A 如果指定字符位於字符串“\AThe”的開頭，則返回匹配項

● \Z 如果指定的字符位於字符串“Spain\Z”的末尾，則返回匹配項

In [16]:
test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\d')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

print()
pattern = re.compile(r'\s')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
    
print()
pattern = re.compile(r'\w')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
    
print()
pattern = re.compile(r'\bhey')
matches = pattern.finditer('heyho hohey') # ho-hey, ho\nhey are matches!
for match in matches:
    print(match)
    
print()
pattern = re.compile(r'\Ahello')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
    
print()
pattern = re.compile(r'123_\Z')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

<re.Match object; span=(6, 7), match='1'>
<re.Match object; span=(7, 8), match='2'>
<re.Match object; span=(8, 9), match='3'>

<re.Match object; span=(5, 6), match=' '>
<re.Match object; span=(10, 11), match=' '>
<re.Match object; span=(16, 17), match=' '>

<re.Match object; span=(0, 1), match='h'>
<re.Match object; span=(1, 2), match='e'>
<re.Match object; span=(2, 3), match='l'>
<re.Match object; span=(3, 4), match='l'>
<re.Match object; span=(4, 5), match='o'>
<re.Match object; span=(6, 7), match='1'>
<re.Match object; span=(7, 8), match='2'>
<re.Match object; span=(8, 9), match='3'>
<re.Match object; span=(9, 10), match='_'>
<re.Match object; span=(11, 12), match='h'>
<re.Match object; span=(12, 13), match='e'>
<re.Match object; span=(13, 14), match='y'>
<re.Match object; span=(14, 15), match='h'>
<re.Match object; span=(15, 16), match='o'>
<re.Match object; span=(17, 18), match='h'>
<re.Match object; span=(18, 19), match='o'>
<re.Match object; span=(19, 20), match='h'>
<re.Match o

<h1><center>Sets</center></h1>
●集合是一對方括號 [] 內的一組具有特殊含義的字符。背靠背附加多個條件，例如 [aA-Z]。

●集合內的 ^（插入符號）否定表達式。

●集合中的 -（破折號）指定一個範圍（如果介於兩者之間），否則為破折號本身。

示例：

- [arn] 返回存在指定字符（a、r 或 n）之一的匹配項 

- [an]  返回任何小寫字符的匹配項，按字母順序在 a 和 n 之間

- [^arn] 返回 a匹配除 a、r 和 n 之外的任何字符

- [0123] 返回存在任何指定數字（0、1、2 或 3）的匹配項 

- [0-9] 返回匹配 0 之間的任何數字和 9

- 0-5返回從 00 到 59 的任何兩位數字的匹配項

- [a-zA-Z] 按字母順序返回 a 和 z 之間的任何字符的匹配項，小寫或大寫

In [18]:
test_string = 'hello 123_'
pattern = re.compile(r'[a-z]')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

<re.Match object; span=(0, 1), match='h'>
<re.Match object; span=(1, 2), match='e'>
<re.Match object; span=(2, 3), match='l'>
<re.Match object; span=(3, 4), match='l'>
<re.Match object; span=(4, 5), match='o'>


In [19]:
dates = '''
01.04.2020

2020.04.01

2020-04-01
2020-05-23
2020-06-11
2020-07-11
2020-08-11

2020/04/02

2020_04_04
2020_04_04
'''

print('all dates with a character in between')
pattern = re.compile(r'\d\d\d\d.\d\d.\d\d')
matches = pattern.finditer(dates)
for match in matches:
    print(match)
print()

print('only dates with - or . in between')
pattern = re.compile(r'\d\d\d\d[-.]\d\d[-.]\d\d') #  no escape for the . here in the set
matches = pattern.finditer(dates)
for match in matches:
    print(match)

print()
print('only dates with - or . in between in May or June')
pattern = re.compile(r'\d\d\d\d[-.]0[56][-.]\d\d')
matches = pattern.finditer(dates)
for match in matches:
    print(match)
    
# a dash in a character set specifies a range if it is in between, otherwise the dash itself
print()
print('only dates with - or . in between in May, June, July')
pattern = re.compile(r'\d\d\d\d[-.]0[5-7][-.]\d\d') #  no escape for the . here in the set
matches = pattern.finditer(dates)
for match in matches:
    print(match)

all dates with a character in between
<re.Match object; span=(13, 23), match='2020.04.01'>
<re.Match object; span=(25, 35), match='2020-04-01'>
<re.Match object; span=(36, 46), match='2020-05-23'>
<re.Match object; span=(47, 57), match='2020-06-11'>
<re.Match object; span=(58, 68), match='2020-07-11'>
<re.Match object; span=(69, 79), match='2020-08-11'>
<re.Match object; span=(81, 91), match='2020/04/02'>
<re.Match object; span=(93, 103), match='2020_04_04'>
<re.Match object; span=(104, 114), match='2020_04_04'>

only dates with - or . in between
<re.Match object; span=(13, 23), match='2020.04.01'>
<re.Match object; span=(25, 35), match='2020-04-01'>
<re.Match object; span=(36, 46), match='2020-05-23'>
<re.Match object; span=(47, 57), match='2020-06-11'>
<re.Match object; span=(58, 68), match='2020-07-11'>
<re.Match object; span=(69, 79), match='2020-08-11'>

only dates with - or . in between in May or June
<re.Match object; span=(36, 46), match='2020-05-23'>
<re.Match object; span=(47

<h1><center>量化</center></h1>
● * : 0 或更多

● + : 1 個或更多

● ? : 0 或 1，當字符可選時使用

● {4} : 確切數字

● {4,6} ：範圍數（最小值，最大值）

In [20]:
my_string = 'hello_123'
pattern = re.compile(r'\d*')
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

print()
pattern = re.compile(r'\d+')
matches = pattern.finditer(my_string)
for match in matches:
    print(match)
    
print()
my_string = 'hello_1_2-3'
pattern = re.compile(r'_?\d')
matches = pattern.finditer(my_string)
for match in matches:
    print(match)
    
print()
my_string = '2020-04-01'
pattern = re.compile(r'\d{4}') # or if you need a range r'\d{3,5}'
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

<re.Match object; span=(0, 0), match=''>
<re.Match object; span=(1, 1), match=''>
<re.Match object; span=(2, 2), match=''>
<re.Match object; span=(3, 3), match=''>
<re.Match object; span=(4, 4), match=''>
<re.Match object; span=(5, 5), match=''>
<re.Match object; span=(6, 9), match='123'>
<re.Match object; span=(9, 9), match=''>

<re.Match object; span=(6, 9), match='123'>

<re.Match object; span=(5, 7), match='_1'>
<re.Match object; span=(7, 9), match='_2'>
<re.Match object; span=(10, 11), match='3'>

<re.Match object; span=(0, 4), match='2020'>


In [21]:
dates = '''
2020.04.01

2020-04-01
2020-05-23
2020-06-11
2020-07-11
2020-08-11

2020/04/02

2020_04_04
2020_04_04
'''
pattern = re.compile(r'\d{4}.\d{2}.\d{2}')
matches = pattern.finditer(dates)
for match in matches:
    print(match)
print()

pattern = re.compile(r'\d+.\d+.\d+')
matches = pattern.finditer(dates)
for match in matches:
    print(match)

<re.Match object; span=(1, 11), match='2020.04.01'>
<re.Match object; span=(13, 23), match='2020-04-01'>
<re.Match object; span=(24, 34), match='2020-05-23'>
<re.Match object; span=(35, 45), match='2020-06-11'>
<re.Match object; span=(46, 56), match='2020-07-11'>
<re.Match object; span=(57, 67), match='2020-08-11'>
<re.Match object; span=(69, 79), match='2020/04/02'>
<re.Match object; span=(81, 91), match='2020_04_04'>
<re.Match object; span=(92, 102), match='2020_04_04'>

<re.Match object; span=(1, 11), match='2020.04.01'>
<re.Match object; span=(13, 23), match='2020-04-01'>
<re.Match object; span=(24, 34), match='2020-05-23'>
<re.Match object; span=(35, 45), match='2020-06-11'>
<re.Match object; span=(46, 56), match='2020-07-11'>
<re.Match object; span=(57, 67), match='2020-08-11'>
<re.Match object; span=(69, 79), match='2020/04/02'>
<re.Match object; span=(81, 91), match='2020_04_04'>
<re.Match object; span=(92, 102), match='2020_04_04'>


<h1><center>條件：Use the "|" for either or condition.</center></h1>


In [22]:
my_string = """
Mr Simpson
Mrs Simpson
Mr. Brown
Ms Smith
Mr. T
"""
pattern = re.compile(r'Mr\.?\s\w+')
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

print()
pattern = re.compile(r'(Mr|Ms|Mrs)\.?\s\w+')
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

<re.Match object; span=(1, 11), match='Mr Simpson'>
<re.Match object; span=(24, 33), match='Mr. Brown'>
<re.Match object; span=(43, 48), match='Mr. T'>

<re.Match object; span=(1, 11), match='Mr Simpson'>
<re.Match object; span=(12, 23), match='Mrs Simpson'>
<re.Match object; span=(24, 33), match='Mr. Brown'>
<re.Match object; span=(34, 42), match='Ms Smith'>
<re.Match object; span=(43, 48), match='Mr. T'>


<h1><center>分組 ( ) is used to group substrings in the matches.</center></h1>

In [23]:
emails = """
pythonengineer@gmail.com
Python-engineer@gmx.de
python-engineer123@my-domain.org
"""
pattern = re.compile('[a-zA-Z1-9-]+@[a-zA-Z-]+\.[a-zA-Z]+')
pattern = re.compile('[a-zA-Z1-9-]+@[a-zA-Z-]+\.(com|de)')
pattern = re.compile('([a-zA-Z1-9-]+)@([a-zA-Z-]+)\.([a-zA-Z]+)')
matches = pattern.finditer(emails)
for match in matches:
    print(match)
    print(match.group(0))
    print(match.group(1))
    print(match.group(2))
    print(match.group(3))

<re.Match object; span=(1, 25), match='pythonengineer@gmail.com'>
pythonengineer@gmail.com
pythonengineer
gmail
com
<re.Match object; span=(26, 48), match='Python-engineer@gmx.de'>
Python-engineer@gmx.de
Python-engineer
gmx
de
<re.Match object; span=(49, 81), match='python-engineer123@my-domain.org'>
python-engineer123@my-domain.org
python-engineer123
my-domain
org


<h1><center>修改字符串</center></h1>
● split(): 將字符串拆分成一個列表，在 RE 匹配的地方拆分它

● sub()：查找RE匹配的所有子字符串，並用不同的字符串替換它們

In [24]:
my_string = 'abc123ABCDEF123abc'
pattern = re.compile(r'123') #  no escape for the . here in the set
matches = pattern.split(my_string)
print(matches)

my_string = "hello world, you are the best world"
pattern = re.compile(r'world')
subbed_string = pattern.sub(r'planet', my_string)
print(subbed_string)

['abc', 'ABCDEF', 'abc']
hello planet, you are the best planet


In [25]:
urls = """
http://python-engineer.com
https://www.python-engineer.org
http://www.pyeng.net
"""
pattern = re.compile(r'https?://(www\.)?(\w|-)+\.\w+')
pattern = re.compile(r'https?://(www\.)?([a-zA-Z-]+)(\.\w+)')
matches = pattern.finditer(urls)
for match in matches:
    #print(match)
    print(match.group()) # 0
    #print(match.group(1))
    #print(match.group(2))
    print(match.group(3))
    
# substitute using back references to replace url + domain name
subbed_urls = pattern.sub(r'\2\3', urls)
print(subbed_urls)

http://python-engineer.com
.com
https://www.python-engineer.org
.org
http://www.pyeng.net
.net

python-engineer.com
python-engineer.org
pyeng.net



<h1><center>Compilation Flags</center></h1>
● ASCII, A ：使 \w、\b、\s 和 \d 等多個轉義符僅匹配具有相應屬性的 ASCII 字符。

● DOTALL，S：製作。匹配任何字符，包括換行符。

● IGNORECASE, I ：進行不區分大小寫的匹配。

● LOCALE, L ：進行區域設置匹配。

● MULTILINE, M : 多行匹配，影響 ^ 和 $。

● VERBOSE, X (for 'extended') : 啟用詳細的 REs，可以更清晰地組織和理解。

In [26]:
my_string = "Hello World"
pattern = re.compile(r'world', re.IGNORECASE) # No match without I flag
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

my_string = '''
hello
cool
Hello
'''
# line starts with ...
pattern = re.compile(r'^[a-z]', re.MULTILINE) # No match without M flag
matches = pattern.finditer(my_string)
for match in matches:
    print(match)

<re.Match object; span=(6, 11), match='World'>
<re.Match object; span=(1, 2), match='h'>
<re.Match object; span=(7, 8), match='c'>
